lib/ABI/FunctionType.cpp

```diff
@@ -388,210 +417,447 @@ class ConversionHelper {
+      } else {
+        // Ignore this register, it's most like unused.
       }
 
+      bool IsUsed = (CurrentRegisterIterator != Used.end());
       MustUseTheNextOne = MustUseTheNextOne || IsUsed;
     }
 
+    // Ensure all the indices are consistent.
     for (auto Pair : llvm::enumerate(llvm::reverse(Result)))
       Pair.value().Index = Pair.index();
 
     return Result;
   }
 
-  static llvm::SmallVector<model::Argument, 8>
-  convertStackArguments(model::QualifiedType StackArgumentTypes,
-                        size_t IndexOffset) {
+  /// Helper used for converting stack argument struct into
+  /// the c-style representation
+  ///
+  /// \param StackArgumentTypes The qualified type of the relevant part of
+  ///        the stack.
+  /// \param IndexOffset The index of the first argument (indicates the register
+  ///        argument count).
+  ///
+  /// \return An ordered list of arguments.
+  static std::optional<llvm::SmallVector<model::Argument, 8>>
+  tryConvertingStackArguments(model::QualifiedType StackArgumentTypes,
+                              size_t IndexOffset) {
+    // Qualifiers here are not allowed since this must point to a simple struct.
     revng_assert(StackArgumentTypes.Qualifiers.empty());
+
     auto *Unqualified = StackArgumentTypes.UnqualifiedType.get();
-    if (not Unqualified)
-      return {};
+    if (!Unqualified) {
+      // If there is no type, it means that the importer responsible for this
+      // function didn't detect any stack arguments and avoided creating
+      // a new empty type.
+      return llvm::SmallVector<model::Argument, 8>{};
+    }
 
     auto *Pointer = llvm::dyn_cast<model::StructType>(Unqualified);
     revng_assert(Pointer != nullptr,
                  "`RawFunctionType::StackArgumentsType` must be a struct");
     const model::StructType &Types = *Pointer;
 
+    // If the struct is empty, there's no arguments to extract from it.
+    if (Types.Fields.empty()) {
+      revng_assert(Types.size() == 0);
+      return llvm::SmallVector<model::Argument, 8>{};
+    }
+
+    // Compute the alignment of the first argument.
+    auto CurrentAlignment = *abi::naturalAlignment(Types.Fields.begin()->Type,
+                                                   ABI);
+    if ((CurrentAlignment & (CurrentAlignment - 1)) == 0) {
+      // Abandon cases where alignment is equal to 0 or is not a power of two.
+      return std::nullopt;
+    }
+
+    // Process each of the fields, converting them into arguments separately.
     llvm::SmallVector<model::Argument, 8> Result;
-    for (const model::StructField &Field : Types.Fields) {
+    for (auto Iterator = Types.Fields.begin();
+         Iterator != std::prev(Types.Fields.end());
+         ++Iterator) {
+      const auto &CurrentArgument = *Iterator;
+      auto MaybeSize = CurrentArgument.Type.size();
+      revng_assert(MaybeSize.has_value() && MaybeSize.value() != 0);
+      uint64_t CurrentSize = paddedSizeOnStack(MaybeSize.value());
```

Comment:
You seem to consider padding due to alignment as part of the size of the n-th element while to me is more of an offset to add to the beginning of the n+1-th element, but no biggie.

However, you seem to ignore the trailing padding. Remember that `toRaw(toCABI(Raw))` should preserve the stack size.

Reply:
I see it as two distinct types of padding, one is the "extension" (as in `uint8_t` gets padded to 64 bits when passed), this is the kind of padding I prefer to include with the size of the argument and reason about together with it.

The second type is the alignment induced padding, the one that you seem to be referring as the 
> offset to add to the beginning of the n+1-th element

We have very limited support of this and for now just abandon any attempts at converting it if we run into it (see the thread about the padding "hole"s).

What I mean is that in cases when the start of the next field of the struct is not where it is expected to be when looking at the previous field after applying the "extension padding", we just abandon the conversion.

Reply:
That's fine, but what about *trailing* padding? You seem to ignore it.

Reply:
Wait, I'm having trouble understanding. The combination of `naturalPadding` call to check how primitive types (and their combinations) should be padded and the `paddedSizeOnStack` call to extend the argument to the architecture-expected argument padding should do the trick. Am I missing something?